File: backend/src/application/use-cases/JoinCommunity.ts
import { User } from '../../domain/entities/User';
import { IUserRepository } from '../../domain/repositories/IUserRepository';
import { ICommunityRepository } from '../../domain/repositories/ICommunityRepository';

export interface JoinCommunityDTO {
  userId: string;
  communityId: string;
}

export class JoinCommunity {
  constructor(
    private readonly userRepository: IUserRepository,
    private readonly communityRepository: ICommunityRepository
  ) {}

  async execute(dto: JoinCommunityDTO): Promise<User> {
    const user = await this.userRepository.findById(dto.userId);
    if (!user) {
      throw new Error('User not found');
    }

    const community = await this.communityRepository.findById(dto.communityId);
    if (!community) {
      throw new Error('Community not found');
    }

    if (!community.isActive) {
      throw new Error('Community is not active');
    }

    user.joinCommunity(dto.communityId);
    const updatedUser = await this.userRepository.update(user);

    community.incrementMemberCount();
    await this.communityRepository.update(community);

    return updatedUser;
  }
}

File: backend-dotnet/src/Hestia.Application/UseCases/JoinCommunity.cs
using Hestia.Domain.Entities;
using Hestia.Domain.Repositories;

namespace Hestia.Application.UseCases;

public class JoinCommunity
{
    private readonly IUserRepository _userRepository;
    private readonly ICommunityRepository _communityRepository;

    public JoinCommunity(IUserRepository userRepository, ICommunityRepository communityRepository)
    {
        _userRepository = userRepository;
        _communityRepository = communityRepository;
    }

    public async Task<User> ExecuteAsync(Guid userId, Guid communityId, CancellationToken ct = default)
    {
        var user = await _userRepository.FindByIdAsync(userId, ct)
            ?? throw new ApplicationException("User not found");

        var community = await _communityRepository.FindByIdAsync(communityId, ct)
            ?? throw new ApplicationException("Community not found");

        if (!community.IsActive)
            throw new ApplicationException("Community is not active");

        user.JoinCommunity(communityId);
        var updated = await _userRepository.UpdateAsync(user, ct);

        community.IncrementMemberCount();
        await _communityRepository.UpdateAsync(community, ct);

        return updated;
    }
}
